### PR TITLE
[SPARK-34360][SQL] Support truncation of v2 tables

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
1. Add new interface `TruncatableTable` which represents tables that allow atomic truncation.
2. Implement new method in `InMemoryTable` and in `InMemoryPartitionTable`.

### Why are the changes needed?
To support `TRUNCATE TABLE` for v2 tables.

### Does this PR introduce _any_ user-facing change?
Should not.

### How was this patch tested?
Added new tests to `TableCatalogSuite` that check truncation of non-partitioned and partitioned tables:
```
$ build/sbt "test:testOnly *TableCatalogSuite"
```

### CODE REVIEW

**Comment by MaxGekk** [comment]
@cloud-fan @HyukjinKwon Could you review this, please.

**Comment by MaxGekk** [comment]
@cloud-fan @HyukjinKwon Any objections to this PR?

**Comment by cloud-fan** [comment]
LGTM, also cc @rdblue @imback82 @dongjoon-hyun

**Review by jaceklaskowski — APPROVED** [review]
LGTM (non-binding)

**Review by imback82 — APPROVED** [review]
LGTM

**Comment by rdblue** [comment]
@MaxGekk, why is this necessary instead of deleting from the table or overwriting everything with no new records? I don't see a good reason to do this, especially at the catalog level instead of the table level. Introducing new ways to do something that is already possible over-complicates the API and is a step in the wrong direction.

Please consider this a -1 until we come to consensus -- I may support it in the end, but I don't want anyone choosing to commit despite disagreement in the mean time.

**Comment by MaxGekk** [comment]
> ... why is this necessary instead of deleting from the table or overwriting everything with no new records?

1. By emulating table truncation via the insertion of no rows, you require atomic operations: delete + insert but a concrete implementation might not support this though it can atomically truncate a table.
2. You close the room for truncation specific optimizations. If a catalog implementation would know in advance that we want to truncate the entire table instead of deleting all rows, it could do that in a more optimal way. Let's say some file based implementation could move the table folder to a trash folder using one atomic syscall.
3. From security or permissions controls point of view, we could distinguish insert with overwrite (or delete) from truncation. I could imagine a case when some roles/users can have only truncation permissions but not insert or delete permissions.
4.  Also it is possible that truncation op is just a record at catalog level log but inserts/deletes are records at table level logs. So, we cannot smoothly sit on such implementation if we emulate table truncation via inserts/deletes.

In general, I do believe we should not hide our intention from catalog implementations - truncation should be explicit. Table catalog implementation should decide how to implement in a more optimal way. So, if they can emulate truncation via overwriting with no rows, ok, this is up to them.

**Comment by rdblue** [comment]
@MaxGekk, can you share the use case that you have for this? You mentioned truncation-specific optimizations. I think working with concrete use cases is usually a good idea. If these are theoretical only -- like a user that can drop all data but not a subset -- then we should put this off. If there's a specific case, then let's discuss it.

I agree that there _may_ be good reason to pass that the engine's intent was to truncate. That's why we have `SupportsTruncate` for the write builder. And I agree with you that we don't necessarily need to use an atomic operation that could truncate and add data at the same time. Your point about not having insert permissions is a good one to justify not using `SupportsTruncate`, although the case of a user that can drop all data but not subsets doesn't sound real. The point about truncation possibly being a metadata operation is why we added `SupportsDelete` at the table level.

Those points may indicate that an interface to truncate a table as a stand-alone operation is valid, although I still think that it is a bad idea to add more interfaces to v2 without a reasonable expectation that they will actually be used.

Another problem here is that this is operation is proposed at the catalog level, which does not fit with how v2 works. I think that the reason for this is emulating what the Hive does, but that's not usually a good choice.

In v2, catalogs load tables and tables are modified. That's why `SupportsDelete` extends `Table` and not `TableCatalog`. This keeps concerns separate, so we have a way to handle tables that don't exist and a separate way to handle tables that don't support a certain operation. Mixing those two together at the catalog level over-complicates the API, requiring a source to throw one exception if the table doesn't exist and another if it doesn't support truncation. (We also went through this discussion with the recently added interfaces to add/drop partitions.)

Assuming that it is worth adding this interface, I would expect it to be a mix-in for `Table`. And like `SupportsOverwrite` that implements `SupportsTruncate`, I think this should also update `SupportsDelete` so that tables don't need to implement both interfaces.

**Comment by MaxGekk** [comment]
> ... can you share the use case that you have for this?

@rdblue For instance, v2 table catalog for JDBC:

0. I assume if a table supports the `SupportsTruncate` interface, it must support atomic remove of all data and write any set of rows (empty and **non-empty**).
1. Support of atomic `SupportsTruncate` is not so easy in the case of JDBC. For example, we still don't support it in JDBC v2 Table Catalog, see SPARK-32595.
2. DBMS usually provides special command for table truncation (see [DB2](https://www.ibm.com/support/knowledgecenter/SSEPEK_10.0.0/sqlref/src/tpc/db2z_sql_truncate.html), [Oracle](https://docs.oracle.com/cd/B28359_01/server.111/b28286/statements_10007.htm#SQLRF01707), [PostgreSQL](https://www.postgresql.org/docs/9.1/sql-truncate.html), [Hive](https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DDL#LanguageManualDDL-TruncateTable)). So, we could map the new method `truncateTable` to a DBMS command.

**Comment by rdblue** [comment]
@MaxGekk, thanks. Then let's work on updating this to fit more cleanly with the design of v2 catalogs and tables. This should be a `Table` interface, not an extension to `TableCatalog`.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TruncatableTable.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+public interface TruncatableTable extends Table {
+  /**
+   * Truncate a table by removing all rows from the table atomically.
+   * If the table supports partitions, the method removes all existing partitions.
+   *
+   * @return true if a table was truncated successfully otherwise false
+   *
+   * @since 3.2.0
+   */
+  boolean truncateTable();
```
Comment (rdblue) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TruncatableTable.java:35]: Why include "table" in the method name? `table.truncate()` seems clear enough to me.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TruncatableTable.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import org.apache.spark.annotation.Evolving;
+
+/**
+ * Represents a table which can be atomically truncated.
+ */
+@Evolving
+public interface TruncatableTable extends Table {
+  /**
+   * Truncate a table by removing all rows from the table atomically.
+   * If the table supports partitions, the method removes all existing partitions.
```
Comment (rdblue) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TruncatableTable.java:29]: Why is this required? A table may support empty partitions that exist independent of rows, like Hive tables. Is there a compelling reason to force this behavior?

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TruncatableTable.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+package org.apache.spark.sql.connector.catalog;
+
+import org.apache.spark.annotation.Evolving;
+
+/**
+ * Represents a table which can be atomically truncated.
+ */
+@Evolving
+public interface TruncatableTable extends Table {
```
Comment (rdblue) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TruncatableTable.java:26]: The other traits use `Supports` as the first word, but `SupportsTruncate` already exists. This name is okay, but others may want to change it.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/connector/InMemoryTable.scala [hunk 4]
```diff
@@ -51,7 +51,7 @@ class InMemoryTable(
     val distribution: Distribution = Distributions.unspecified(),
     val ordering: Array[SortOrder] = Array.empty)
   extends Table with SupportsRead with SupportsWrite with SupportsDelete
-      with SupportsMetadataColumns {
+      with SupportsMetadataColumns with TruncatableTable {
```
Comment (rdblue) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/connector/InMemoryTable.scala:54]: I think that `SupportsDelete` should implement `TruncatableTable`, similar to how `SupportsOverwrite` implements `SupportsTruncate`. That way it isn't necessary to for implementations to support both delete and truncate separately.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TruncatableTable.java (hunk 3 above)
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TruncatableTable.java:26]: Maybe, do we need `Atomic` in the name? For example, `SupportsAtomicTruncate` is better?

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TruncatableTable.java (hunk 2 above)
Comment (rdblue) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TruncatableTable.java:29]: `SupportsPartitionManagement` has a `truncatePartition` method that will "Truncate a partition in the table by completely removing partition data."

That conflicts with the behavior of truncate here, which drops partitions. I think this requirement should be removed.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TruncatableTable.java (hunk 3 above)
Comment (rdblue) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TruncatableTable.java:26]: All of the operations should be atomic, right? Certainly, that's the expectation of a write that truncates and then appends data.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TruncatableTable.java (hunk 1 above)
Comment (rdblue) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TruncatableTable.java:35]: Actually, this seems to mirror the other uses so I'm okay with it. I'd probably remove it but it seems okay to leave as is if you feel strongly about it.

**Comment by cloud-fan** [comment]
I like the idea to allow the catalog implementations to distinguish between `TRUNCATE TABLE t` and `INSERT OVERWRITE t SELECT * FROM empty_table`. It seems fine to move the API to the table side. But the name conflict is a bit annoying.

One idea is to keep the name `SupportsTruncate`, and put it in `org.apache.spark.sql.connector.catalog`, which is the same package of `SupportsDelete`. Then there is no conflict.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TruncatableTable.java (hunk 2 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TruncatableTable.java:29]: > Is there a compelling reason to force this behavior?

Yes, I wanted to highlight that it can remove not only all rows but also partitions to align to Spark's v1 (and Hive) `TRUNCATE TABLE` but the command doesn't drop partitions in fact:
```sql
spark-sql> CREATE TABLE tbl (col0 INT) PARTITIONED BY (part INT);
spark-sql> INSERT INTO tbl PARTITION (part=0) SELECT 0;
spark-sql> ALTER TABLE tbl ADD PARTITION (part=1);
spark-sql> SHOW PARTITIONS tbl;
part=0
part=1
spark-sql> SELECT * FROM tbl;
0	0
spark-sql> TRUNCATE TABLE tbl;
spark-sql> SHOW PARTITIONS tbl;
part=0
part=1
spark-sql> SELECT * FROM tbl;
spark-sql>
```
Let me remove the sentence.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TruncatableTable.java (hunk 1 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TruncatableTable.java:35]: I just followed naming convention (maybe informal one) in other interfaces. For instance, if a table implements `TruncatableTable`, `SupportsPartitionManagement` and `SupportsAtomicPartitionManagement`, we have 3 methods in the same namespace:
- truncatePartition()
- truncatePartitions()
- and truncate()

In that case, maybe it is better to name this method as `truncateTable()` which can highlight that it is applied to entire table.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TruncatableTable.java (hunk 2 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TruncatableTable.java:29]: but in this PR, I drop empty partitions as well. The question is should v2 `TRUNCATE TABLE` be aligned to v1 implementation, and preserve empty partitions? I guess the answer is yes.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/connector/InMemoryTable.scala (hunk 4 above)
Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/connector/InMemoryTable.scala:54]: If you are sure that if an implementation supports `SupportsDelete ` then it must support `TruncatableTable` too. But I slightly doubt about it. I could imagine a case when an implementation can delete rows by a filter but cannot guarantee atomic truncation.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TruncatableTable.java (hunk 3 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TruncatableTable.java:26]: if we name this interface as `SupportsAtomicTruncate`, someone may guess that `SupportsTruncate` is opposite to it (means non-atomic truncate) or both atomic and non-atomic truncate. But actually, those two interfaces are "orthogonals" in some sense.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TruncatableTable.java (hunk 2 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TruncatableTable.java]: Surprising but we don't have any tests that checks partition existence after entire table truncation. I opened this PR https://github.com/apache/spark/pull/31544 with such checks for tables from v1 In-Memory and Hive external catalogs.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/connector/InMemoryTable.scala (hunk 4 above)
Comment (rdblue) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/connector/InMemoryTable.scala:54]: Truncate is equivalent to `deleteWhere(true)`. Why would that not be equivalent?

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TruncatableTable.java (hunk 3 above)
Comment (rdblue) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TruncatableTable.java:26]: I don't think that adding `Atomic` to the name is a good idea. The other operations don't specify whether an operation is atomic and I don't think that this should necessarily either. If I understand correctly, the purpose of this is to allow using `TRUNCATE` for JDBC or similar optimizations. That's more likely to be atomic, but may not be. It looks like the Hive implementation would not be because the partitions are kept.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java [hunk 5]
```diff
@@ -68,4 +69,14 @@ default boolean canDeleteWhere(Filter[] filters) {
    * @throws IllegalArgumentException If the delete is rejected due to required effort
    */
   void deleteWhere(Filter[] filters);
+
+  @Override
+  default boolean truncateTable() {
+    Filter[] filters = new Filter[] { new AlwaysTrue() };
```
Comment (rdblue) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:75]: Could you make this a constant instead of creating a new filter and array instance here?

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Looks good now. Thanks for fixing this, @MaxGekk!

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java [hunk 6]
```diff
@@ -68,4 +69,15 @@ default boolean canDeleteWhere(Filter[] filters) {
    * @throws IllegalArgumentException If the delete is rejected due to required effort
    */
   void deleteWhere(Filter[] filters);
+
+  Filter[] ALWAYS_TRUE_FILTER = new Filter[] { new AlwaysTrue() };
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:73]: This becomes a public API as well. Shall we put it in an internal object like `CatalogV2Util`?

Comment (MaxGekk) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:73]: I wouldn't do that because of:
- I believe interfaces should be independent from internals as much as possible.
- ALWAYS_TRUE_FILTER can be used in other methods of the interface like `deleteWhere()`. For example, when an implementation overrides `truncateTable()` but an user want to delete all rows via `deleteWhere()`, so, he/she can re-use the constant.
- Other interfaces have constants too, for example https://github.com/apache/spark/blob/ec1560af251d2c3580f5bccfabc750f1c7af09df/sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TableCatalog.java#L47 . I don't see much difference between `PROP_LOCATION` and  `ALWAYS_TRUE_FILTER`.

Comment (HyukjinKwon) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:73]: Can't people use `AlwaysTrue` directly if they want? I would also prefer to avoid having multiple ways to do the same thing. It looks odd to have the AlwaysTrue predicate constant as an API under `SupportsDelete` interface.

Comment (MaxGekk) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:73]: How about to revert this commit https://github.com/apache/spark/pull/31475/commits/d1e5a18066f9fb2ff0ca1504e7c3f0802905febd , and implement it as:
```scala
default boolean truncateTable() {
    Filter[] filters = new Filter[] { new AlwaysTrue() };
   ...
```
I am not sure it is worth to do this premature optimization. Comparing to the truncation op, the allocation overhead is small. If it is a hot spot, JVM should do all work for us and optimize it, I do believe.

Comment (HyukjinKwon) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:73]: Okay, I just noticed that it was changed per the review comment above. I think it's fine to remove as the overhead is small, and avoid exposing `ALWAYS_TRUE_FILTER` as an API.

**Review by HyukjinKwon — APPROVED** [review]
Looks fine to me too.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by HyukjinKwon** [comment]
Merged to master.
